src/v/cloud_storage/types.h

```diff
@@ -420,43 +420,78 @@ struct anomalies
 
 std::ostream& operator<<(std::ostream& o, const anomalies& a);
 
-enum class upload_object_type {
+enum class upload_type {
     object,
     segment_index,
     manifest,
     group_offsets_snapshot,
     download_result_file,
     remote_lifecycle_marker,
 };
+std::ostream& operator<<(std::ostream&, upload_type);
 
-std::ostream& operator<<(std::ostream&, upload_object_type);
+enum class download_type {
+    object,
+    segment_index,
+};
+std::ostream& operator<<(std::ostream&, download_type);
 
-class remote_probe;
+template<typename T>
+struct transferred_type_traits;
 
-// Represents a request to upload an object as a payload, enables
-// aggregate initialization by the caller.
-struct upload_object_request {
-    using probe_callback_t = ss::noncopyable_function<void(remote_probe&)>;
+template<>
+struct transferred_type_traits<upload_type> {
+    static constexpr auto default_value = upload_type::object;
+    using buf_t = iobuf;
+};
 
-    cloud_storage_clients::bucket_name bucket_name;
+template<>
+struct transferred_type_traits<download_type> {
+    static constexpr auto default_value = download_type::object;
+    using buf_t = iobuf&;
+};
+
+template<typename transfer_kind>
+concept is_object_transfer = requires {
+    transferred_type_traits<transfer_kind>::default_value;
+    typename transferred_type_traits<transfer_kind>::buf_t;
+};
+
+class remote_probe;
+using probe_callback_t = ss::noncopyable_function<void(remote_probe&)>;
+
+template<typename transferred_object_type>
+requires is_object_transfer<transferred_object_type>
+struct object_transfer_request {
+    cloud_storage_clients::bucket_name bucket;
     cloud_storage_clients::object_key key;
 
-    // The payload of the object to upload
-    iobuf payload;
+    transferred_type_traits<transferred_object_type>::buf_t payload;
     retry_chain_node& parent_rtc;
 
-    upload_object_type upload_type{upload_object_type::object};
+    transferred_object_type transferred_type{
+      transferred_type_traits<transferred_object_type>::default_value};
```

Comment:
To remove the template parameter on this request, we could use a variant here and rely on default initialization, eg 

```
using transfer_t = std::variant<upload_type, download_type>;
transfer_t transferred_type{};
```

Similarly we could use a variant for the buffer type:
```
std::variant<iobuf, std::reference_wrapper<iobuf>>;
```

But now we just have the one request type, this would allow the caller to call `upload_object` with `download_object_request`, iobuf& with `upload_type` etc. With the template parameter and associated traits it is not possible to use the `upload_object` or `download_object` with an incorrect combination of request, iobuf type or object type.

Reply:
>this would allow the caller to call upload_object with download_object_request

Do you have a particular use case for this in mind? IMO this is a good reason to _not_ overload the payload, and instead encapsulate some transfer-agnostic request args (bucket, key, etc), and separately have the download or upload payload. It otherwise seems like it'd be error-prone and harder to verify code by inspection (e.g. conceptually it seems wrong to implicitly allow uploading with a download request -- would it always be a bug? if so, why allow it?)

Something like:

```
struct object_request_params {
    bucket_name bucket;
    object_key key;
    retry_chain_node& parent_rtc;
    ....
};

struct upload_request {
    object_request_params params;
    iobuf payload;
};

struct download_request {
    object_request_params params;
    iobuf& payload;
};
```

Reply:
My comment was ambiguous, sorry. To be clear, as the PR is implemented now, an upload request is _not_ usable with a download object method, and a download request is similarly not usable with an upload object method. Right now the types ensure that only the right request can be used by the right method.

If there had been a simple `transfer_request` object, there would have been a possibility of such a mixup which we avoid right now.

That said, I like the suggestion of extracting a struct with the common bits and creating the request objects by composing that and the upload/download specific fields. It looks like it might be much simpler and easier to read, I will try it out in this PR, thanks

Reply:
I tried this change and the definition is much simpler:

```
struct transfer_object_details {
    cloud_storage_clients::bucket_name bucket;
    cloud_storage_clients::object_key key;
    retry_chain_node& parent_rtc;

    std::optional<probe_callback_t> success_cb{std::nullopt};
    std::optional<probe_callback_t> failure_cb{std::nullopt};
    std::optional<probe_callback_t> backoff_cb{std::nullopt};

    void on_success(remote_probe& probe);
    void on_failure(remote_probe& probe);
    void on_backoff(remote_probe& probe);
};

struct upload_object_request {
    transfer_object_details transfer;
    iobuf payload;
    upload_type type{upload_type::object};
};

struct download_object_request {
    transfer_object_details transfer;
    iobuf& payload;
    download_type type{download_type::object};
};
```

The usage is more complicated:

```
co_await _remote.upload_object({
          .transfer = {
            .bucket = _conf->bucket_name,
            .key = cloud_storage_clients::object_key{index_path},
            .parent_rtc = fib,
            .success_cb = [](auto& probe) { probe.index_upload(); },
            .failure_cb = [](auto& probe) { probe.failed_index_upload(); }},
          .payload = idx_res->index.to_iobuf(),
          .type = cloud_storage::upload_type::segment_index,
        });
```

I'd prefer the simpler usage as there are many call sites.

Reply:
I may be missing something, but the call sites don't seem significantly more complicated than in the other approach, compared to the complexity of the `object_transfer_request`?

I also still don't really understand why the types of the download and upload request should the same given they're doing fundamentally different things -- I get that the enum and variant gives us this distinction, but it seems weird to lump them together.

Reply:
I think the nested initializers are harder to read at a glance than flat initializers, eg

```
remote.local()
        .upload_object({
          .transfer_details
          = {.bucket = bucket, .key = path, .parent_rtc = fib},
          .payload = make_iobuf_from_string("p"),
        })
```

vs

```
remote.local()
        .upload_object({.bucket = bucket, .key = path, .parent_rtc = fib,
          .payload = make_iobuf_from_string("p")
        })
```

But I don't feel strongly about this and readability is subjective, so I have made the changes.

Reply:
Thank you! Very much appreciate the accommodation.